### ISSUE TITLE
Filename ops on Sequenceable Collection

### ISSUE THREAD

**Issue by Beechside** [supercollider/supercollider#3324]
I manipulate a lot of files and often want to process all files in a folder.  Since `basename` and `splitext` don't work on a sequenceable collection of fully qualified filenames, I have the following in my extensions:

```
+SequenceableCollection {
	basename {^this.performUnaryOp(\basename)}
	splitext {^this.performUnaryOp(\splitext)}
}
```

However, I wondered if there was any reason why it should not be added to the main branch.  There may be others in pathname support and if the two above are viable candidates, I'm happy to go through them all and see what works and what doesn't.

Thoughts?

**Comment by telephon** [comment]
There have been many of such cases, and indeed often we have decided to extend the collection classes. The main argument for me is whether code becomes more universal and better to read. I'm not sure if this is really a good case for that, because file processing usually happens in rather preparational bits of code. Writing `paths.collect(_.basename)` isn't that much longer than `paths.basename`. But I don't have a strong opinion here.

On the long run, we should introduce a `Lift` (or similar) class, so that you can wrap a collection in such a way that it will always distribute commands. You'd then write: `path = pathMatch(...).lift` and then you could just operate on all paths as if you were operating on a single path. The advantage is that this applies to all use cases.

I've prepared an implmentation for that, but we need to see how much needed it really is (see https://github.com/telephon/Neutral)

**Comment by jamshark70** [comment]
Potentially, one could argue that any method of any class could be added to the Array interface. We have to draw a line somewhere. Maybe the line should include these basic path methods, maybe not -- I'm not sure. But, once we start drawing the line more permissively, it gets harder to tighten interfaces down again. So I'd suggest caution. It feels a bit like bloat to me -- not that it wouldn't be convenient, but in the past, we've shot ourselves in the foot by adding conveniences hastily and then realizing later that the convenience introduced other problems.

Is this really a unary operator, btw? I'm not sure of the precedent for considering string methods to be in the same category as numeric operators. If it's added at all, I'd suggest to use `collect` here. I find more and more that a direct approach may be more boring than a cleverly magical solution, but it's better long term for maintenance.

If it's annoying to type `collect` often, another alternative is the ddwSnippets quark.

```
DDWSnippets.put("base", ".collect(_.basename)");

DDWSnippets.learn;  // set a hotkey
```

**Comment by Beechside** [comment]
Thank you both.  I encapsulate metadata in my file names and need to extract/analyze the filename (excluding filetype).  Hence, without extensions, I end up coding something like:

`allFiles.collect{ |file| file.basename.splitext[0]}`

since partial application doesn't work.  With extensions, this is coded:

`allFiles.basename.splitext.slice(nil, 0)`

It may be, as Telephon hints, less easy to read.  Interestingly, `.bench` shows the extensions to be slower than `.collect`, which offends my compulsive behaviour regarding optimisation, whether important or not :-)

I still think that both `.basename` and `.splitext` should make the cut.  I haven't experienced the problems you describe so will yield but I found myself having to dig into the source code when they didn't work.  It seems unintuitive not to support arrays.

**Comment by brianlheim** [comment]
> The main argument for me is whether code becomes more universal and better to read. I'm not sure if this is really a good case for that, because file processing usually happens in rather preparational bits of code.

I agree with @telephon here.

There are two problems w/r/t universality - if adopted, these would be the only string methods supported in this way by the SequenceableCollection interface – not counting, for instance, `asInteger`, which is coincidentally there for math support. It doesn't take much to extend the reasoning here to argue for adding more methods from the String interface. This is problematic because many methods supported by String (`rotate`, `asSymbol`, `scramble`, to name a few) are also supported by SequenceableCollection. So, if we expanded the number of String-interface methods beyond the two proposed here, I can see users being often confused about whether or not they can get away with calling a certain string-supported method on a collection of strings.

In short, either we end up with a minimal reimplementation of the String interface, or a full-as-possible reimplementation that buys back the typing it saves through time spent making new mistakes. I don't think that either of these design results are suitable for the core lib.

While we're on the topic, I would discourage the use of strings to represent paths, in favor of `PathName`, which will be more robust for cross-platform work. In this case you have `PathName:-fileNameWithoutExtension` for partial application.

**Comment by telephon** [comment]
just a note @beechside, to avoid troubles with partial application,you can just  wrap your own function first: `f = {|x| x.basename.splitext.first }`.

**Comment by Beechside** [comment]
> While we're on the topic, I would discourage the use of strings to represent paths, in favor of `PathName`

Yes, how true!  If only `.pathMatch` returned class of PathName or there were a `.asPathName` method or if `.as` worked for PathName and one could do either of these on an array of strings. 

Whoops!  Back to the same argument... ;o)

**Comment by telephon** [comment]
for the fun of it, why not try this?

```
Quarks.install("Neutral");
```

and then you just write:

```supercollider
 "~/*".pathMatch.lift.basename.splitext.first.unlift
```

I'm curious of good applications for that system, because one day it could become standard.

**Comment by Beechside** [comment]
Interesting - After you mentioned lift, I checked the Haskell implementation.  Definitely worth a try...

However: 

(1) `Quarks.install("Neutral");` doesn't work.  I had to specify the fully-qualified path 
`Quarks.install("https://github.com/telephon/Neutral");`

(2) Despite SC announcing that Neutral was installed there's nothing in either system or user downloaded-quarks.  Should there be?  Consequently, the lift method was not recognised.

Will be very happy to try it out if you can point me in the right direction...

(PS: I tried on both 3.8 and 3.9dev)

**Comment by telephon** [comment]
@Beechside did you recompile the class library?

>Quarks.install("Neutral"); doesn't work.

hm, might need to open a separate issue if you continue to have this problem. You need to have git installed.

**Comment by telephon** [comment]
(done now).

**Comment by brianlheim** [comment]
Can we close this based on consensus vote -> no?

**Comment by Beechside** [comment]
Well, I guess I've been voted down!  I'm still not sure why one would not extend the collection classes but I don't maintain the code...

Thank you, @telephon - Lift works well enough but I would argue it's pretty unintuitive in this case.  Certainly less intuitive than using basename and splitext on a collection.  I realise it may have other application more sophisticated than my simple example, however and look forward it it joining the main branch.  Its performance is also better than I had expected and it outperforms basename & splitext on collections.

I would also be quite happy to use PathName more rigorously, as suggested by @brianlheim but I might argue it has limitations (not used in pathMatch and also not supported in collections).

I appreciate your comments and the work you put into the system.

**Comment by telephon** [comment]
In the end, if you don't plan to distribute your code as code snippets (rather than a library), then it's totally fine to have your own class extensions, it's what we all have!

> I'm still not sure why one would not extend the collection classes but I don't maintain the code...

The reason is that there is no systematic reason, but only a specific practical reason, and it seems so far nobody else really feels the need for it. If many turn up with the same, this might change.

**Comment by jamshark70** [comment]
I think there is a systematic reason: math operators are more general than string methods for paths, and they are part of a broadly supported interface in a way that path-related methods are not.

In `a + b`, the operands may be:

- numbers
- sequenceable collections
- numeric wrappers (Complex, Polar etc)
- geometry objects (Point, Rect)
- patterns
- UGens
- functions (any subclass of AbstractFunction)

Array:+ by itself doesn't necessarily make sense.

Array:+ as part of a system-wide pattern of broad support for math operators in at least a dozen relevant object types... that's a design decision (which James McCartney implemented brilliantly).

As convenient as it would be, there's no way that `basename` can reach the same breadth of relevance.

**Comment by Beechside** [comment]
Very helpful responses.  Going back to my original question, I encapsulate metadata within the filename and I'm keen to find out how others might do that so I'll revert to the forums, I think, since it's more to do with practice than sclang...

**Comment by telephon** [comment]
the easiest way to encapsulate different data is to use a Dictionary, or one of its subclasses like IdentityDictionary.

**Comment by Beechside** [comment]
Thank as always but, in this case, I'm thinking of something external to SC.  I have images, MIDI files, PDF, notation, multiple sound files and data that I need to group together.  Am looking at both SQL and noSQL solutions.  Probably have to do the latter and maybe something like CouchDB, though I'm currently trawling through Sourceforge to see what others have done...
